File: src/com/peyo/lbsettings/DetailsActivity.java
package com.peyo.lbsettings;

import android.app.Activity;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import androidx.leanback.app.DetailsFragment;
import androidx.leanback.widget.AbstractDetailsDescriptionPresenter;
import androidx.leanback.widget.Action;
import androidx.leanback.widget.ArrayObjectAdapter;
import androidx.leanback.widget.ClassPresenterSelector;
import androidx.leanback.widget.DetailsOverviewRow;
import androidx.leanback.widget.FullWidthDetailsOverviewRowPresenter;
import androidx.leanback.widget.HeaderItem;
import androidx.leanback.widget.ListRow;
import androidx.leanback.widget.ListRowPresenter;
import androidx.leanback.widget.OnItemViewClickedListener;
import androidx.leanback.widget.Presenter;
import androidx.leanback.widget.Row;
import androidx.leanback.widget.RowPresenter;

import com.google.gson.Gson;

import java.io.IOException;
import java.io.InputStream;

public class DetailsActivity extends Activity {
	@Override
	protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_details);
	}

	public static class MenuDetailsFragment extends DetailsFragment {
		int titleRes;
		int iconRes;
		@Override
		public void onCreate(Bundle savedInstanceState) {
			super.onCreate(savedInstanceState);
			titleRes = getActivity().getIntent().getIntExtra("TitleRes", 0);
			iconRes = getActivity().getIntent().getIntExtra("IconRes", 0);

			ClassPresenterSelector selector = new ClassPresenterSelector();
			selector.addClassPresenter(DetailsOverviewRow.class,
					new FullWidthDetailsOverviewRowPresenter(
							new DetailsDescriptionPresenter()));
			selector.addClassPresenter(ListRow.class, new ListRowPresenter());

			ArrayObjectAdapter rows = new ArrayObjectAdapter(selector);
			rows.add(getOverview());
			rows.add(getRow());
			setAdapter(rows);

			setOnItemViewClickedListener(new OnItemViewClickedListener() {
				@Override
				public void onItemClicked(Presenter.ViewHolder vh, Object item,
										  RowPresenter.ViewHolder rowViewHolder, Row row) {
					if (item instanceof Action) {
						Action action = (Action) item;
						if (action.getId() == 1) {
							Intent intent = new Intent(getContext(), GuidedStepActivity.class);
							intent.putExtra("TitleRes", titleRes);
							intent.putExtra("IconRes", iconRes);
							startActivity(intent);
						}
					} else if (item instanceof MenuItem) {
						Intent intent = new Intent(getContext(), DetailsActivity.class);
						MenuItem menuItem = (MenuItem) item;
						intent.putExtra("IconRes", menuItem.getIconRes(getContext()));
						intent.putExtra("TitleRes", menuItem.getTitleRes(getContext()));
						startActivity(intent);
					}
					getActivity().finish();
				}
			});
		}

		public class DetailsDescriptionPresenter extends AbstractDetailsDescriptionPresenter {
			@Override
			protected void onBindDescription(
					AbstractDetailsDescriptionPresenter.ViewHolder vh, Object item) {
				vh.getTitle().setText(getResources().getString(titleRes));
				vh.getSubtitle().setText("Menu Item Details");
				vh.getBody().setText("Lorem ipsum dolor sit amet, consectetur "
						+ "adipisicing elit, sed do eiusmod tempor incididunt ut labore "
						+ " et dolore magna aliqua. Ut enim ad minim veniam, quis "
						+ "nostrud exercitation ullamco laboris nisi ut aliquip ex ea "
						+ "commodo consequat.");
			}
		}

		DetailsOverviewRow getOverview() {
			DetailsOverviewRow overview = new DetailsOverviewRow("Menu Item Details");

			Drawable drawable = getResources().getDrawable(iconRes);
			overview.setImageDrawable(drawable);

			overview.addAction(new Action(1, "OK"));
			overview.addAction(new Action(2, "Cancel"));

			return overview;
		}

		ListRow getRow() {
			String json = inputStreamToString(getResources().openRawResource(R.raw.personal));
			MenuItem[] items =  new Gson().fromJson(json, MenuItem[].class);

			ArrayObjectAdapter row = new ArrayObjectAdapter(new MenuItemPresenter());
			for (MenuItem item : items) {
				row.add(item);
			}

			return new ListRow(new HeaderItem(getResources().getString(R.string.personal)), row);
		}

		String inputStreamToString(InputStream inputStream) {
			try {
				byte[] bytes = new byte[inputStream.available()];
				inputStream.read(bytes, 0, bytes.length);
				return new String(bytes);
			} catch (IOException e) {
				return null;
			}
		}
	}

}

File: src/com/peyo/lbsettings/DetailsActivity.kt
package com.peyo.lbsettings

import android.content.Intent
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.leanback.app.DetailsSupportFragment
import androidx.leanback.widget.AbstractDetailsDescriptionPresenter
import androidx.leanback.widget.Action
import androidx.leanback.widget.ArrayObjectAdapter
import androidx.leanback.widget.ClassPresenterSelector
import androidx.leanback.widget.DetailsOverviewRow
import androidx.leanback.widget.FullWidthDetailsOverviewRowPresenter
import androidx.leanback.widget.HeaderItem
import androidx.leanback.widget.ListRow
import androidx.leanback.widget.ListRowPresenter
import androidx.leanback.widget.OnItemViewClickedListener

import com.google.gson.Gson

import java.io.IOException
import java.io.InputStream

class DetailsActivity : FragmentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_details)
    }

    class MenuDetailsFragment : DetailsSupportFragment() {
        private var titleRes: Int = 0
        private var iconRes: Int = 0

        private val overview: DetailsOverviewRow
            get() {
                val adapter = ArrayObjectAdapter()
                adapter.add(Action(1, "OK"))
                adapter.add(Action(2, "Cancel"))

                val overview = DetailsOverviewRow("Menu Item Details")
                overview.actionsAdapter = adapter
                overview.imageDrawable = resources.getDrawable(iconRes, null)

                return overview
            }

        private val row: ListRow
            get() {
                val json = inputStreamToString(resources.openRawResource(R.raw.personal))
                val items = Gson().fromJson(json, Array<MenuItem>::class.java)

                val row = ArrayObjectAdapter(MenuItemPresenter())
                for (item in items) {
                    row.add(item)
                }

                return ListRow(HeaderItem(resources.getString(R.string.personal)), row)
            }

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            titleRes = activity!!.intent.getIntExtra("TitleRes", 0)
            iconRes = activity!!.intent.getIntExtra("IconRes", 0)

            val selector = ClassPresenterSelector()
            selector.addClassPresenter(DetailsOverviewRow::class.java,
                    FullWidthDetailsOverviewRowPresenter(
                            DetailsDescriptionPresenter()))
            selector.addClassPresenter(ListRow::class.java, ListRowPresenter())

            val rows = ArrayObjectAdapter(selector)
            rows.add(overview)
            rows.add(row)
            adapter = rows

            onItemViewClickedListener = OnItemViewClickedListener { _, item, _, _ ->
                if (item is Action) {
                    if (item.id == 1L) {
                        val intent = Intent(context, GuidedStepActivity::class.java)
                        intent.putExtra("TitleRes", titleRes)
                        intent.putExtra("IconRes", iconRes)
                        startActivity(intent)
                    }
                } else if (item is MenuItem) {
                    val intent = Intent(context, DetailsActivity::class.java)
                    intent.putExtra("IconRes", item.getIconRes(context!!))
                    intent.putExtra("TitleRes", item.getTitleRes(context!!))
                    startActivity(intent)
                }
                activity!!.finish()
            }
        }

        inner class DetailsDescriptionPresenter : AbstractDetailsDescriptionPresenter() {
            override fun onBindDescription(
                    vh: AbstractDetailsDescriptionPresenter.ViewHolder, item: Any) {
                vh.title.text = resources.getString(titleRes)
                vh.subtitle.text = "Menu Item Details"
                vh.body.text = ("Lorem ipsum dolor sit amet, consectetur "
                        + "adipisicing elit, sed do eiusmod tempor incididunt ut labore "
                        + " et dolore magna aliqua. Ut enim ad minim veniam, quis "
                        + "nostrud exercitation ullamco laboris nisi ut aliquip ex ea "
                        + "commodo consequat.")
            }
        }

        private fun inputStreamToString(inputStream: InputStream): String? {
            try {
                val bytes = ByteArray(inputStream.available())
                inputStream.read(bytes, 0, bytes.size)
                return String(bytes)
            } catch (e: IOException) {
                return null
            }

        }
    }

}
